tuf/ngclient/updater.py

```diff
@@ -0,0 +1,575 @@
+
+                # This exception will occur if the target does not exist
+                # locally.
+                except sslib_exceptions.StorageError:
+                    updated_targets.append(target)
+                    updated_targetpaths.append(target_filepath)
+                    break
+
+                # The file does exist locally, check if its hash differs.
+                if digest_object.hexdigest() != digest:
+                    updated_targets.append(target)
+                    updated_targetpaths.append(target_filepath)
+                    break
+
+        return updated_targets
+
+    def download_target(
+        self,
+        targetinfo: Dict,
+        destination_directory: str,
+        target_base_url: Optional[str] = None,
+    ):
+        """
+        Download target specified by 'targetinfo' into 'destination_directory'.
+
+        Args:
+            targetinfo: data received from get_one_valid_targetinfo()
+            destination_directory: existing local directory to download into.
+                Note that new directories may be created inside
+                destination_directory as required.
+            target_base_url: Optional; Base URL used to form the final target
+                download URL. Default is the value provided in Updater()
+
+        Raises:
+            TODO: download-related errors
+            TODO: file write errors
+        """
+        if target_base_url is None and self._target_base_url is None:
+            raise ValueError(
+                "target_base_url must be set in either download_target() or "
+                "constructor"
+            )
+        if target_base_url is None:
+            target_base_url = self._target_base_url
+        else:
+            target_base_url = _ensure_trailing_slash(target_base_url)
+
+        full_url = parse.urljoin(target_base_url, targetinfo["filepath"])
+
+        with download.download_file(
+            full_url, targetinfo["fileinfo"].length, self._fetcher
+        ) as target_file:
+            _check_file_length(target_file, targetinfo["fileinfo"].length)
+            _check_hashes_obj(target_file, targetinfo["fileinfo"].hashes)
+
+            filepath = os.path.join(
+                destination_directory, targetinfo["filepath"]
+            )
+            sslib_util.persist_temp_file(target_file, filepath)
+
+    def _download_metadata(
+        self, rolename: str, length: int, version: Optional[int] = None
+    ) -> bytes:
+        """Download a metadata file and return it as bytes"""
+        if version is None:
+            filename = f"{rolename}.json"
+        else:
+            filename = f"{version}.{rolename}.json"
+        url = parse.urljoin(self._metadata_base_url, filename)
+        return download.download_bytes(
+            url,
+            length,
+            self._fetcher,
+            strict_required_length=False,
+        )
+
+    def _load_local_metadata(self, rolename: str) -> bytes:
```

Comment:
This feels [familiar](https://github.com/trishankatdatadog/tuf-on-a-plane/blob/ee8049a63ea439d72509f6e876eef08fb50c6ab1/src/tuf_on_a_plane/readers.py#L60) :)